### PR TITLE
Generate training set via Offline Store

### PR DESCRIPTION
**Pull request by simba-git**

### CODE REVIEW

**Review by sdreyer — DISMISSED** [review]
Nothing major that stands out, a lot to go through though

File: ff/serving/provider/offline.go [hunk 1]
```diff
@@ -106,8 +351,39 @@ func newMemoryOfflineTable() *memoryOfflineTable {
...
+			if i == 0 {
+				return nil
+			}
+			return sortedRecs[i-1].Value
+		} else if i == lastIdx {
+			// Every record happened before the label.
+			return rec.Value
+		}
+	}
+	panic("Unable to getLastValue before timestamp")
```
Comment (sdreyer) [inline ff/serving/provider/offline.go:382]: Do we want this to panic or should it just return an error?

File: ff/serving/provider/offline.go [hunk 2]
```diff
@@ -89,13 +170,177 @@ func (store *memoryOfflineStore) CreateResourceTable(id ResourceID) (OfflineTabl
...
+	Label    interface{}
+}
+
+type trainingRowsIterator struct {
+	data trainingRows
+	idx  int
+}
+
+func newTrainingRowsIterator(data trainingRows) TrainingSetIterator {
+	return &trainingRowsIterator{
```
Comment (Sami1309) [inline ff/serving/provider/offline.go:317]: perhaps call this memoryTrainingSetIterator to be more specific on the implementation

File: ff/serving/provider/offline.go [hunk 3]
```diff
@@ -89,13 +170,177 @@ func (store *memoryOfflineStore) CreateResourceTable(id ResourceID) (OfflineTabl
...
+			return err
+		}
+		features[i] = feature
+	}
+	labelRecs := label.records()
+	trainingData := make([]trainingRow, len(labelRecs))
+	for i, rec := range labelRecs {
+		featureVals := make([]interface{}, len(features))
+		for i, feature := range features {
+			featureVals[i] = feature.getLastValueBefore(rec.Entity, rec.TS)
```
Comment (Sami1309) [inline ff/serving/provider/offline.go:269]: getlastvaluebefore returns nil if no entitymap to the given entity is set, but panics if it is but there are no resources. It should return nil in both cases I think

File: ff/serving/provider/offline.go (hunk 1 above)
Comment (Sami1309) [reply ff/serving/provider/offline.go:382]: or return nil

File: ff/serving/provider/offline_test.go [hunk 4]
```diff
@@ -44,18 +58,24 @@ func TestOfflineStores(t *testing.T) {
...
 }
 
-func randomResourceID() ResourceID {
-	possibleTypes := []OfflineResourceType{Label, Feature}
-	randType := possibleTypes[rand.Intn(2)]
+func randomID(types ...OfflineResourceType) ResourceID {
+	var t OfflineResourceType
+	if len(types) == 0 {
+		t = NoType
+	} else if len(types) == 1 {
```
Comment (Sami1309) [inline ff/serving/provider/offline_test.go:65]: else if and else can be combined since they perform the same logic, t = types[rand.Intn(len(types))]

**Review by Sami1309 — APPROVED** [review]
very minor changes, still fine and lgtm tho

File: ff/serving/provider/offline.go (hunk 1 above)
Comment (simba-git) [reply ff/serving/provider/offline.go:382]: We never get to this line, but I can't return nothing so it is a panic

File: ff/serving/provider/offline.go (hunk 3 above)
Comment (simba-git) [reply ff/serving/provider/offline.go:269]: Ill clarify the panic in a comment

File: ff/serving/provider/offline.go (hunk 2 above)
Comment (simba-git) [reply ff/serving/provider/offline.go:317]: good call

File: ff/serving/provider/offline_test.go (hunk 4 above)
Comment (simba-git) [reply ff/serving/provider/offline_test.go:65]: true, but I'd argue it a little clearer this way

**Pull request merged**